test/adaptive_load/adaptive_load_controller_test.cc

```diff
@@ -0,0 +1,847 @@
+#include <chrono>
+
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/registry/registry.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/input_variable_setter.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/config/utility.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/benchmark_result.pb.h"
+#include "api/adaptive_load/input_variable_setter_impl.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/adaptive_load/metrics_plugin_impl.pb.h"
+#include "api/adaptive_load/scoring_function_impl.pb.h"
+#include "api/adaptive_load/step_controller_impl.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.grpc.pb.h"
+#include "api/client/service.pb.h"
+#include "api/client/service_mock.grpc.pb.h"
+
+#include "grpcpp/test/mock_stream.h"
+
+#include "test/adaptive_load/fake_plugins/fake_metrics_plugin/fake_metrics_plugin.h"
+#include "test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.h"
+#include "test/adaptive_load/fake_time_source.h"
+#include "test/adaptive_load/minimal_output.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_join.h"
+#include "adaptive_load/metrics_plugin_impl.h"
+#include "adaptive_load/plugin_loader.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using ::testing::_;
+using ::testing::DoAll;
+using ::testing::Eq;
+using ::testing::HasSubstr;
+using ::testing::Return;
+using ::testing::SetArgPointee;
+
+/**
+ * Creates a valid TypedExtensionConfig proto selecting the real BinaryScoringFunction plugin
+ * and configuring it with a threshold.
+ *
+ * @param lower_threshold Threshold value to set within the config proto.
+ *
+ * @return TypedExtensionConfig Full scoring function plugin spec that selects
+ * nighthawk.binary-scoring and provides a config.
+ */
+envoy::config::core::v3::TypedExtensionConfig
+MakeLowerThresholdBinaryScoringFunctionConfig(double lower_threshold) {
```

Comment:
`const double lower_treshold`?

Reply:
Shouldn't need that as it's passed by value

Reply:
ah yes, sorry about that.